lib/stages/VisSharedMemWriter.cpp

```diff
@@ -0,0 +1,379 @@
+#include <system_error>         // for system_error
+#include <tuple>                // for get
+#include <utility>              // for pair
+#include <vector>               // for vector
+#include "Hash.hpp"             // for Hash
+#include "StageFactory.hpp"     // for REGISTER_KOTEKAN_STAGE, StageMakerTemplate
+#include "datasetManager.hpp"   // for datasetManager
+#include "datasetState.hpp"     // for freqState, eigenvalueState, inputState, prodState, stackState
+#include "fmt.hpp"              // for format, fmt
+#include "kotekanLogging.hpp"   // for INFO, DEBUG, ERROR, FATAL_ERROR
+#include "visBuffer.hpp"        // for visFrameView, visMetadata
+#include "visUtil.hpp"          // for time_ctype, frameID, ts_to_double
+
+using kotekan::bufferContainer;
+using kotekan::Config;
+using kotekan::Stage;
+
+
+REGISTER_KOTEKAN_STAGE(VisSharedMemWriter);
+
+VisSharedMemWriter::VisSharedMemWriter(Config& config, const std::string& unique_name, bufferContainer& buffer_container) :
+    Stage(config, unique_name, buffer_container, std::bind(&VisSharedMemWriter::main_thread, this)) {
+
+        // Fetch any simple configuration
+        _root_path = config.get_default<std::string>(unique_name, "root_path", "/dev/shm/");
+        _sem_name = config.get_default<std::string>(unique_name, "sem_name", "kotekan");
+        _fname_buf = config.get_default<std::string>(unique_name, "fname_buf", "calBuffer");
+        _ntime = config.get_default<size_t>(unique_name, "nsamples", 512);
+
+        // Setup the input vector
+        in_buf = get_buffer("in_buf");
+        register_consumer(in_buf, unique_name.c_str());
+
+        // Check if any of the old buffer files exist
+        // Remove them, if they do
+        DEBUG("Checking for and removing old buffer files...");
+        check_remove(_root_path + "sem." + _sem_name);
+        check_remove(_root_path + _fname_buf);
+
+
+}
+
+VisSharedMemWriter::~VisSharedMemWriter() {
+    // make sure to unlink the semaphore and unmap the mappings
+    // We are setting num_writes to 0 in the structured data,
+    // to communicate to readers that the ring buffer is not being written to
+
+    wait_for_semaphore();
+    num_writes = 0;
+    memcpy(structured_data_addr, &num_writes, sizeof(num_writes));
+
+    if (sem_post(sem) == -1) {
+        FATAL_ERROR("Failed to release semaphore {}", _sem_name);
+        return;
+    }
+}
+
+bool VisSharedMemWriter::wait_for_semaphore() {
+    // handles timed waits for semaphores
+    // does a standard wait if the system clock is not accessible
+
+    struct timespec ts;
+    if (clock_gettime(CLOCK_REALTIME, &ts) == -1){
+        WARN("Failed to get system time. {:d} ({:s})\n Not using timed semaphores.", errno, std::strerror(errno));
+        if (sem_wait(sem) == -1) {
+            ERROR("Failed to acquire semaphore {}", _sem_name);
+            return false;
+        }
+        return true;
+    }
+
+    ts.tv_sec += 2;
```

Comment:
Hi! It is most definitely arbitrary. I asked Rick what would be a reasonable amount, and he suggested 1 second. I then bumped that to 2 seconds. 

I think it should definitely be configurable, that would make modifying it not require an additional deployment. I will make that adjustment. I think it would be hard to determine what it should be without running tests where readers are reading the buffer as kotekan is writing to it.

The behaviour depends on at which point it fails to claim the lock. If the frame has not been writing to the buffer yet, it drops the frame, and moves to the next one. If the frame had been written to the buffer, but that update has not been recorded to the access record yet, the access record does not get updated, and that frame may as well "be considered dropped". 

Do you have an alternative suggested behaviour?

Reply:
I am currently cleaning up all of the logic that determines timeout behaviour, to ensure it is properly logged and consistent.

Reply:
To be honest, it's not obvious to me that your should have a time wait at all.

Data comes in at ~100 frames / second, although you'll probably get it within one ~5s burst every 10s. Let's say that the reader blocks things for 10s. That means that I'm going to take a frame, wait for 2s, drop it and then repeat for 5 frames. At the same time I have data backing up in the system which is going to back up for 1000 frames all the way up to the network receive and cause *all* incoming data to get dropped.

So in that sense your timed wait is too long, to prevent a scenario like this you need to make it be around (1/(inbound rate)) ~10ms. However, it's very plausible that the reader might block for that long, and there is enough buffer depth (and the processing rate fast enough), that a 10ms slow down doesn't matter. In that sense, 10ms is too short. 

In all I think this argument suggests that dropping data based on this timed rate isn't the right thing to be doing.

I would suggest doing this:

- Using the timed wait with a really long timeout (more like 2 minutes), and if you wait that long throw a FATAL_ERROR.
- Using prometheus to export the length of time that you wait to acquire the semaphore. Then we can alert of dangerous slow downs.
- Adding a Valve stage before your buffer process in the config to prevent thing backing up.

Reply:
It would be good to get @nritsche's input here too.

Reply:
That work has been pushed.

Reply:
Oh, I had an out-dated page opened and have not seen your comments!

Reply:
This make sense. And the valve stage provides a prometheus metric, right?

Reply:
Yup! We'll need to ensure there's a large buffer depth in front of it, but we have a boat load of memory on recv2 as well now, so that's easy.